kinesis/kinesis.go

```diff
@@ -637,6 +646,79 @@ func gzipCompress(data []byte) ([]byte, error) {
 	return b.Bytes(), nil
 }
 
+func compressThenTruncate(compressorFunc CompressorFunc, data []byte, maxOutLen int, truncatedSuffix []byte, outputPlugin OutputPlugin) ([]byte, error) {
+
+	truncationReductionPercent := 90
```

Comment:
why the percentage is 90?
is this needed to be configurable?
Or could be a const defined in the const section at the very beginning of this file.

Reply:
I second this- and also how did you determine that this algorithm and percentage is ideal in the sense that it makes some trade off between efficiency and minimizing data loss? Some sort of mini-design review here might be worthwhile.

Reply:
If we keep as a constant, I can move it to the beginning of the file.
Ah I see. Yes, 90% is arbitrary padding to account for noise in the randomness of the logs before compression. As discussed prior to implementation, Wesley proposed giving some padding to the size before compression, to value underestimate, over another iteration. Not sure if 90% is the optimal number.

This 10% reduction seemed to me like a decent number for padding, but it was arbitrarily picked.

If we want to go with a more general solution, which could actually be less prone to abuse, we could instead of having a constant 90%, have something like 100 - 5%, 100 - 10%, 100 - 20%, 100 - 40%, 100 - 80%, failure.
That way there are a max 6 iterations, rather than a large number in some cases, which could allow for a smart dos attack.... At first, I wanted to make the algo more straight forward, but it might be a good idea.

I can work on a document for this.

Reply:
Thanks and I think recording your thinking here is very worthwhile. And my challenge to you is to try to work on making it both accurate and informative but also short and to the point.

Reply:
Please see: https://quip-amazon.com/bbHPAwzI3Kwg/Compression-Truncation